File: frontend/src/api/calendar-sync.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiFetch, hasToken } from './client'

export interface CalendarConnection {
  id: string
  user_id: string
  household_id: string
  name: string
  provider: string
  server_url: string
  calendar_url: string
  username: string
  color: string
  sync_token: string | null
  last_synced_at: string | null
  enabled: boolean
  error_count: number
  last_error: string | null
  created_at: string
}

export interface CalendarConnectionCreate {
  name: string
  provider?: string
  server_url: string
  calendar_url: string
  username: string
  password: string
  color?: string
}

export interface CalendarConnectionUpdate {
  name?: string
  color?: string
  enabled?: boolean
  password?: string
}

export interface ExternalEventOccurrence {
  id: string
  connection_id: string
  title: string
  description: string | null
  start_time: string
  end_time: string
  all_day: boolean
  location: string | null
  source_calendar_name: string
  source_calendar_color: string
  provider: string
}

export interface FeedToken {
  token: string
  url: string
}

export function useCalendarConnections() {
  return useQuery({
    queryKey: ['calendar-connections'],
    queryFn: () => apiFetch<CalendarConnection[]>('/calendar/connections'),
    enabled: hasToken(),
  })
}

export function useCreateCalendarConnection() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: CalendarConnectionCreate) =>
      apiFetch<CalendarConnection>('/calendar/connections', {
        method: 'POST',
        body: JSON.stringify(data),
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['calendar-connections'] })
      qc.invalidateQueries({ queryKey: ['external-events'] })
    },
  })
}

export function useUpdateCalendarConnection() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ connectionId, ...data }: CalendarConnectionUpdate & { connectionId: string }) =>
      apiFetch<CalendarConnection>(`/calendar/connections/${connectionId}`, {
        method: 'PATCH',
        body: JSON.stringify(data),
      }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['calendar-connections'] }),
  })
}

export function useDeleteCalendarConnection() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (connectionId: string) =>
      apiFetch<void>(`/calendar/connections/${connectionId}`, { method: 'DELETE' }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['calendar-connections'] })
      qc.invalidateQueries({ queryKey: ['external-events'] })
    },
  })
}

export function useTriggerSync() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (connectionId: string) =>
      apiFetch<CalendarConnection>(`/calendar/connections/${connectionId}/sync`, {
        method: 'POST',
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['calendar-connections'] })
      qc.invalidateQueries({ queryKey: ['external-events'] })
    },
  })
}

export function useExternalEvents(householdId: string, start: string, end: string) {
  return useQuery({
    queryKey: ['external-events', householdId, start, end],
    queryFn: () =>
      apiFetch<ExternalEventOccurrence[]>(
        `/households/${householdId}/external-events?start=${start}&end=${end}`
      ),
    enabled: !!householdId && hasToken(),
  })
}

export function useFeedToken() {
  return useQuery({
    queryKey: ['feed-token'],
    queryFn: () => apiFetch<FeedToken>('/calendar/feed-token'),
    enabled: hasToken(),
  })
}

export function useRegenerateFeedToken() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: () =>
      apiFetch<FeedToken>('/calendar/feed-token/regenerate', { method: 'POST' }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['feed-token'] }),
  })
}

File: backend/app/schemas/calendar_sync.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


class CalendarConnectionCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    provider: Literal["icloud", "nextcloud", "caldav"] = "caldav"
    server_url: str = Field(min_length=1, max_length=2000)
    calendar_url: str = Field(min_length=1, max_length=2000)
    username: str = Field(min_length=1, max_length=500)
    password: str = Field(min_length=1, max_length=500)
    color: str = Field(default="#6C5CE7", max_length=7)

    @field_validator("server_url", "calendar_url")
    @classmethod
    def validate_https(cls, v: str) -> str:
        if not v.startswith("https://"):
            raise ValueError("URL must use HTTPS")
        return v

    @field_validator("color")
    @classmethod
    def validate_color(cls, v: str) -> str:
        if not v.startswith("#") or len(v) != 7:
            raise ValueError("Color must be a hex color like #6C5CE7")
        return v


class CalendarConnectionUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    color: str | None = Field(default=None, max_length=7)
    enabled: bool | None = None
    password: str | None = Field(default=None, min_length=1, max_length=500)

    @field_validator("color")
    @classmethod
    def validate_color(cls, v: str | None) -> str | None:
        if v is not None and (not v.startswith("#") or len(v) != 7):
            raise ValueError("Color must be a hex color like #6C5CE7")
        return v


class CalendarConnectionResponse(BaseModel):
    id: str
    user_id: str
    household_id: str
    name: str
    provider: str
    server_url: str
    calendar_url: str
    username: str
    color: str
    sync_token: str | None
    last_synced_at: datetime | None
    enabled: bool
    error_count: int
    last_error: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class ExternalEventResponse(BaseModel):
    id: str
    connection_id: str
    title: str
    description: str | None
    start_time: datetime
    end_time: datetime
    all_day: bool
    location: str | None
    source_calendar_name: str
    source_calendar_color: str
    provider: str

    model_config = {"from_attributes": True}


class FeedTokenResponse(BaseModel):
    token: str
    url: str
